### PR TITLE
feat: Wasm storage for Data Request & Overlay binaries

### LINKED ISSUE #41: ✨Wasm storage for Data Request & Overlay binaries
## ✨ Feature

Replace the skeletal module `x/storage` to create a new module called `x/wasm-storage` for storing all Wasm binaries used in the SEDA protocol. 

## Implementation

This new module's store will contain two separate keys for storing Data Request Wasm binaries and Overlay Wasm binaries. 

The Wasm binaries keyed by their hashes will be stored in compressed bytecode format along with the specific type information. Note that Data Request Wasm binaries are further categorized into Data Request Wasm and Tally Wasm, while Overlay Wasm binaries are further categorized into Data Request Executor Wasm and Relayer Wasm.

```
0x00 | Hash(DataRequestWasm) -> Wasm{}
0x01 | Hash(OverlayWasm) -> Wasm{}
```

The module's transactions for storing Wasms will look as follows:
```
$ seda-chaind tx wasm-storage store-data-request-wasm <filename> --wasm-type data-request
$ seda-chaind tx wasm-storage store-data-request-wasm <filename> --wasm-type tally

$ seda-chaind tx wasm-storage store-overlay-wasm <filename> --wasm-type data-request-executor
$ seda-chaind tx wasm-storage store-overlay-wasm <filename> --wasm-type relayer
```

Make sure to include ways to query a binary using its hash and list hashes of all binaries. 
```
$ seda-chaind query wasm-storage data-request-wasm <wasm_hash>
$ seda-chaind query wasm-storage overlay-wasm <wasm_hash>

$ seda-chaind query wasm-storage all-data-request-wasms
$ seda-chaind query wasm-storage all-overlay-wasms
```

### PR DESCRIPTION
**Pull request by hacheigriega**
## ✨ Feature

The new `x/wasm-storage` module provides a place where Data Request Issuers can store their Data Request Wasm binaries, which can then be fetched by Overlay nodes. It also stores Overlay Wasm binaries in a separate bucket. Another PR will put these Overlay Wasms under governance control.

## Implementation

As discussed in the issue (#41), there are two separate buckets for Data Request Wasms and Overlay Wasms.
```
0x00 | Hash(DataRequestWasm) -> Wasm{}
0x01 | Hash(OverlayWasm) -> Wasm{}
```

where `Wasm{}` struct type looks as follows:
```
type Wasm struct {
	Hash     []byte
	Bytecode []byte
	WasmType WasmType
}
```

Transactions on this module looks as follows:
```
$ seda-chaind tx wasm-storage store-data-request-wasm <filename> --wasm-type data-request
$ seda-chaind tx wasm-storage store-data-request-wasm <filename> --wasm-type tally

$ seda-chaind tx wasm-storage store-overlay-wasm <filename> --wasm-type data-request-executor
$ seda-chaind tx wasm-storage store-overlay-wasm <filename> --wasm-type relayer
```

Queries can be made using hashes provided in transaction results. There are also endpoints for listing all hashes and Wasm types in the store:
```
$ seda-chaind query wasm-storage data-request-wasm <hash>
$ seda-chaind query wasm-storage overlay-wasm <hash>


$ seda-chaind query wasm-storage list-data-request-wasms
$ seda-chaind query wasm-storage list-overlay-wasms
```

Please note that gas adjustment flag may have to be set to about 1.2 since simulation often underestimates gas usage. That is, add the flags `--gas auto --gas-adjustment 1.2` to transactions. Also note that following the CosmWasm standard we set the Wasm size limit at 819.2 KB in its uncompressed state. This limit is also used during unzip process to prevent zip bombing.

## Testing

Testing will be addressed in a separate PR. Checked that transaction and query endpoints work as expected with Wasm files of varying sizes.

## Related PRs and Issues

Closes #41

### CODE REVIEW

File: proto/sedachain/wasm_storage/query.proto
Comment (gluax) [inline proto/sedachain/wasm_storage/query.proto:38]: Possibly a dumb question:

What's the `yaml` specifier for here in these messages? Isn't `json` more standard for `rpc` things?

File: x/wasm-storage/types/wasm.go
Comment (gluax) [inline x/wasm-storage/types/wasm.go:26]: Ignore this marker. It's so we know what hash algo to use on the overlay side :).

File: x/wasm-storage/keeper/msg_server.go
Comment (gluax) [inline x/wasm-storage/keeper/msg_server.go:31]: I think the goal is to be able to overwrite a hash since we are permissionless yes @mariocao @FranklinWaller?

File: x/wasm-storage/types/keys.go
Comment (gluax) [inline x/wasm-storage/types/keys.go:9]: It may be a stylistic choice on my end, but I'm unsure if others would agree; however, I noticed we have some configured constants in many different files. I think it's easier to find/edit if they are usually in the same place, like a `consts.go` file, and then labeled appropriately in there. Food for thought, not a hard ask.

**Review by gluax — COMMENTED** [review]
It looks good to me, waiting on approval though to double-check the overwriting part :).

File: x/wasm-storage/types/wasm.go
Comment (FranklinWaller) [inline x/wasm-storage/types/wasm.go:33]: Maybe for future usage we can add the block it was added to? So we don't have to migrate our state if we ever introduce pruning

File: x/wasm-storage/types/wasm.go
Comment (FranklinWaller) [inline x/wasm-storage/types/wasm.go:11]: I'm not sure so correct me if i'm wrong but isn't this better to have as a genesis config thing? So we can always upgrade it through a gov decision

File: x/wasm-storage/keeper/msg_server.go
Comment (FranklinWaller) [inline x/wasm-storage/keeper/msg_server.go:79]: I don't remember if i asked this question earlier but is this safe against zipbomb attacks?

File: x/wasm-storage/keeper/msg_server.go
Comment (FranklinWaller) [reply x/wasm-storage/keeper/msg_server.go:31]: The hash is based on the bytes so overwriting is not a thing right

Comment (hacheigriega) [reply x/wasm-storage/keeper/msg_server.go:31]: @FranklinWaller Oh the `Wasm` struct has the field `WasmType`, which could be overwritten.

Comment (hacheigriega) [reply x/wasm-storage/keeper/msg_server.go:31]: We could just get rid of `WasmType` if it's unnecessary to store this info on chain.

File: x/wasm-storage/types/keys.go
Comment (hacheigriega) [reply x/wasm-storage/types/keys.go:9]: That's a reasonable suggestion, but I'm just following the standards of Cosmos projects here. (putting all key-related stuff in `keys.go`)

File: x/wasm-storage/types/wasm.go
Comment (hacheigriega) [reply x/wasm-storage/types/wasm.go:33]: I agree. Not sure if we're going with block number-based vs block time-based pruning but I'll add block number for now.

File: x/wasm-storage/types/wasm.go
Comment (hacheigriega) [reply x/wasm-storage/types/wasm.go:11]: Yea that was actually my initial idea but got lazy 😋

File: x/wasm-storage/keeper/msg_server.go
Comment (hacheigriega) [reply x/wasm-storage/keeper/msg_server.go:79]: Yes I checked that. `MaxWasmSize` should prevent that.

File: x/wasm-storage/keeper/msg_server.go
Comment (gluax) [reply x/wasm-storage/keeper/msg_server.go:31]: Yea, that's the only thing I was wondering about. I should have been more specific. My bad!

Eventually, if/when we add metadata too, that could be an issue... maybe rather than storing metadata and that stuff on the chain, it could emit an event or something for the indexer to grab with that info? i.e. we still take it as args, but we don't store it just give it to indexer in some way.

File: proto/sedachain/wasm_storage/query.proto
Comment (hacheigriega) [reply proto/sedachain/wasm_storage/query.proto:38]: Not dumb at all.. probably for query results? Copy-pasting went too far there. I removed them.

File: x/wasm-storage/types/wasm.go
Comment (hacheigriega) [reply x/wasm-storage/types/wasm.go:33]: Acutally I added timestamp of when the object was added because I think pruning should be based on timestamps (not important anyways right now)

Comment (FranklinWaller) [reply x/wasm-storage/types/wasm.go:36]: To be honest, i think block number is probably better because it's more consistent. Timestamps are not really to be trusted. But maybe it doesn't matter in this use case. We can resolve this one for now.

File: x/wasm-storage/types/wasm.go
Comment (FranklinWaller) [reply x/wasm-storage/types/wasm.go]: Hahaha, no worries, atleast it's noted

**Review by FranklinWaller — APPROVED** [review]

**Comment by hacheigriega** [comment]
@gluax I'll merge if you approve :)

File: x/wasm-storage/keeper/msg_server.go
Comment (Thomasvdam) [reply x/wasm-storage/keeper/msg_server.go:31]: Late to the party but I think the idea is to have the metadata be indexer only. So as long as you have the private key used to upload the bytes you can edit the metadata through the explorer.

**Pull request merged**
